win32ss/user/ntuser/ghost.c

```diff
@@ -176,20 +442,85 @@ BOOL FASTCALL IntMakeHungWindowGhosted(HWND hwndHung)
 
     if (IntIsGhostWindow(pHungWnd))
     {
-        DPRINT1("IntIsGhostWindow\n");
+        ERR("IntIsGhostWindow (hwndHung: %p)\n", (void *)hwndHung);
         return FALSE;   // ghost window cannot be ghosted
     }
 
-    if (IntGhostWindowFromHungWindow(pHungWnd))
+    // "Shell_TrayWnd" is a special window for Taskbar.
+    if (IntIsShellTrayWnd(pHungWnd))
     {
-        DPRINT("Already ghosting\n");
-        return FALSE;   // already ghosting
+        TRACE("IntIsShellTrayWnd (hwndHung: %p)\n", (void *)hwndHung);
+        return FALSE;   // Shell_TrayWnd
+    }
+
+    // "Progman" is a special window for Desktop.
+    if (IntIsProgmanWnd(pHungWnd))
+    {
+        TRACE("IntIsProgmanWnd (hwndHung: %p)\n", (void *)hwndHung);
+        return FALSE;   // Progman
     }
 
-    // TODO: Find a way to pass the hwnd of pHungWnd to the ghost thread as we can't pass parameters directly
+    // create ghost window and add to list
+    pEntry = IntCreateGhostAndAddToList(hwndHung);
+    if (!pEntry)
+    {
+        ERR("IntCreateGhostAndAddToList failed: %p\n", (void *)hwndHung);
+        return FALSE;
+    }
 
     if (!gptiGhostThread)
-        UserCreateSystemThread(ST_GHOST_THREAD);
+    {
+        if (gpGhostStartupEvent == NULL)
+        {
+            // create event
+            gpGhostStartupEvent = ExAllocatePoolWithTag(NonPagedPool,
+                                                        sizeof(KEVENT),
+                                                        USERTAG_GHOST);
+            if (!gpGhostStartupEvent)
+            {
+                ERR("Unable to allocate event\n");
+                return FALSE;   // creation failed
+            }
+        }
+        KeInitializeEvent(gpGhostStartupEvent, NotificationEvent, FALSE);
+
+        // create ghost thread
+        if (!UserCreateSystemThread(ST_GHOST_THREAD))
+        {
+            ERR("Unable to create ghost thread for pHungWnd(%p), hwndHung(%p)\n",
+                (void *)pHungWnd, (void *)hwndHung);
+            IntDeleteGhostEntry(hwndHung);
+            return FALSE;   // creation failed
+        }
+
+        // wait for event
+        UserLeave();
+        Status = KeWaitForSingleObject(gpGhostStartupEvent, UserRequest, UserMode, FALSE, NULL);
+        UserEnterExclusive();
+
+        if (Status == STATUS_USER_APC)
+        {
+            ERR("KeWaitForSingleObject failed: Status: 0x%08lx\n", Status);
+            IntDeleteGhostEntry(hwndHung);
+            return FALSE;
+        }
+
+        // delete gpGhostStartupEvent
+        ExFreePoolWithTag(gpGhostStartupEvent, USERTAG_GHOST);
```

Comment:
My mistake, sorry for the back and forth -- but it now occurs to me that two threads can get here. So this (and the assignment to NULL in the next line) was better at the end of the ghost thread, near `gptiGhostThread = NULL;`

The rest of the thread synchronization stuff looks safe to me now. Woo!

Reply:
OK.